### PR TITLE
Cryk UI news service

### PR DESCRIPTION
**Pull request by CipDan**

### CODE REVIEW

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    formState: { isSubmitSuccessful, errors },
+  } = useForm<NewsCardInput>({
+    resolver: zodResolver(newsCardSchema),
+  });
+
+  useEffect(() => {
+    if (isSubmitSuccessful) {
+      reset();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (DianaCrainic) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: remove comment

Comment (IoanSava) [reply cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: Why did you disable this eslint rule?

File: cryk-ui/src/components/AlertDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import DialogContentText from '@mui/material/DialogContentText';
+import DialogTitle from '@mui/material/DialogTitle';
+import { DocumentNode } from 'graphql';
+import { useMutation } from '@apollo/client';
+import "./css/AlertDialog.css";
+import { GetPaginatedCryptoNewsInput } from '../models/GetPaginatedCryptoNewsInput';
+
+
+export default function AlertDialog(props: {id: string, alertQuery: DocumentNode, refetchQuery: DocumentNode, refetchVars: GetPaginatedCryptoNewsInput, queryEndpoint: string}) {
+  const [open, setOpen] = React.useState(false);
```
Comment (IoanSava) [inline cryk-ui/src/components/AlertDialog.tsx]: rename it to `isOpen` and `setIsOpen`
https://dev.to/michi/tips-on-naming-boolean-variables-cleaner-code-35ig

File: cryk-ui/src/components/AlertDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  const handleClickOpen = () => {
+    setOpen(true);
+  };
+
+  const handleClose = () => {
+    setOpen(false);
+  };
+
+  const handleAgree = (e: React.MouseEvent) => {
```
Comment (IoanSava) [inline cryk-ui/src/components/AlertDialog.tsx]: `e` => `event`

File: cryk-ui/src/components/AlertDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
```
Comment (IoanSava) [inline cryk-ui/src/components/AlertDialog.tsx]: These comments apply for each file:
* use double quotes everywhere
* let's format

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      required_error: "News card title is required"
+    }).min(1, { message: "News card title is required" }),
+    body: string({
+      required_error: "News card body is required"
+    }).min(1, { message: "News card body is required" })
+});
+
+type NewsCardInput = TypeOf<typeof newsCardSchema>;
+
+const Transition = React.forwardRef(function Transition(
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: Maybe we can use an anonymous function: 
```javascript
const Transition = React.forwardRef(function (...
```

File: cryk-ui/src/components/AlertDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import { GetPaginatedCryptoNewsInput } from '../models/GetPaginatedCryptoNewsInput';
+import { RefetchInput } from '../models/RefetchInput';
+import { GetPaginatedCryptocurrenciesInput } from '../models/GetPaginatedCryptocurrenciesInput';
+
+
+export default function AlertDialog(props: {id: string, alertQuery: DocumentNode, refetchInput: RefetchInput<GetPaginatedCryptoNewsInput|GetPaginatedCryptocurrenciesInput>}) {
+  const [open, setOpen] = React.useState(false);
+  const refetchInput = props.refetchInput
+
+  const [ removeNewsEntry ] = useMutation(props.alertQuery, {
```
Comment (IoanSava) [inline cryk-ui/src/components/AlertDialog.tsx]: rename it to `removeEntry`

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+import { UpdateCryptoNewsInput } from "../models/UpdateCryptoNewsInput";
+import { useMutation } from "@apollo/client";
+import "./css/CreateUpdateNewsCardDialog.css";
+import { GetPaginatedCryptoNewsInput } from "../models/GetPaginatedCryptoNewsInput";
+import { GetPaginatedCryptocurrenciesInput } from "../models/GetPaginatedCryptocurrenciesInput";
+import { RefetchInput } from "../models/RefetchInput";
+
+const newsCardSchema = object({
+    title: string({
+      required_error: "News card title is required"
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: Is it restricted to snake case? Or we can use camel case (`requiredError`)?

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  props: TransitionProps & {
+    children: React.ReactElement;
+  },
+  ref: React.Ref<unknown>,
+) {
+  return <Slide direction="up" ref={ref} {...props} />;
+});
+  
+export default function CreateUpdateNewsCardDialog(props: {operationType: string, dialogQuery: DocumentNode, refetchInput: RefetchInput<GetPaginatedCryptoNewsInput|GetPaginatedCryptocurrenciesInput>, cryptocurrencyId: string, news?: News}) {
+  const [open, setOpen] = React.useState(false);
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: rename it to `isOpen` and `setIsOpen`

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+const Transition = React.forwardRef(function Transition(
+  props: TransitionProps & {
+    children: React.ReactElement;
+  },
+  ref: React.Ref<unknown>,
+) {
+  return <Slide direction="up" ref={ref} {...props} />;
+});
+  
+export default function CreateUpdateNewsCardDialog(props: {operationType: string, dialogQuery: DocumentNode, refetchInput: RefetchInput<GetPaginatedCryptoNewsInput|GetPaginatedCryptocurrenciesInput>, cryptocurrencyId: string, news?: News}) {
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: Do we want to make it reusable? I think we should only use this component for news creation/update.

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    context: {clientName: refetchInput.context},
+    refetchQueries: [ {query: refetchInput.query, context: {clientName: refetchInput.context}, variables: refetchInput.variables} ]
+  });
+
+  const onSubmitHandler: SubmitHandler<NewsCardInput> = (values) => {
+    if (props.operationType === "create"){
+      const operationInput: CreateCryptoNewsInput = {title: values.title, body: values.body, about: [ props.cryptocurrencyId ]};
+      createUpdateNewsEntry({
+        variables: {createCryptoNewsInput: operationInput}
+      }).catch((e) => {console.log(JSON.stringify(e, null, 2))});
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: `e` => `error`

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    else {
+      const operationInput: UpdateCryptoNewsInput = {id: (props.news ? props.news.id : ""), title: values.title, body: values.body};
+      createUpdateNewsEntry({
+        variables: {updateCryptoNewsInput: operationInput}
+      }).catch((e) => {console.log(JSON.stringify(e, null, 2))});
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: `e` => `error`

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+              edge="start"
+              color="inherit"
+              onClick={handleClose}
+              aria-label="close"
+              size="medium"
+            >
+              <CloseIcon fontSize="large"/>
+            </IconButton>
+            <DialogTitle className="news-dialog-title">
+              {props.operationType === "create" ? "Create News Card" : "Update News Card"}
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: Let's create a map containing these values. Afterwards we can refactor these ternary operators to something like this:
```javascript
{textValuesMap[props.operationType]["dialog-title"]}
```

File: cryk-ui/src/index.js [hunk 13]
```diff
@@ -2,10 +2,22 @@ import "./index.css";
 import React from "react";
 import ReactDOM from "react-dom";
 import App from "./App";
-import { ApolloClient, ApolloProvider, InMemoryCache } from "@apollo/client";
+import { ApolloClient, ApolloLink, ApolloProvider, HttpLink, InMemoryCache } from "@apollo/client";
+
+const endpoint1 = new HttpLink({
```
Comment (IoanSava) [inline cryk-ui/src/index.js]: Let's rename it to `cryptocurrenciesGraphqlEndpoint`

File: cryk-ui/src/index.js [hunk 14]
```diff
@@ -2,10 +2,22 @@ import "./index.css";
...
+  uri: "http://localhost:3000/graphql"
+})
+
+const endpoint2 = new HttpLink ({
```
Comment (IoanSava) [inline cryk-ui/src/index.js]: Let's rename it to `newsGraphqlEndpoint`

File: cryk-ui/src/index.js [hunk 15]
```diff
@@ -2,10 +2,22 @@ import "./index.css";
...
+  uri: "http://localhost:4000/graphql"
+})
 
 const graphqlClient = new ApolloClient({
-  uri: "http://localhost:3000/graphql",
+  link: ApolloLink.split(
+    operation => operation.getContext().clientName === 'endpoint2',
```
Comment (IoanSava) [inline cryk-ui/src/index.js]: Use `newsGraphqlEndpoint` name

File: cryk-ui/src/models/GetPaginatedCryptoNewsInput.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+export interface GetPaginatedCryptoNewsInput {
+    cryptocurrencyId: string,
```
Comment (IoanSava) [inline cryk-ui/src/models/GetPaginatedCryptoNewsInput.ts]: Use `;` at the end of each property of each model

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 17]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
     }
 `;
 
+const NEWS_PER_PAGE = 2;
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Let's display at least 5 news

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 18]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
+            body
+        }
+    }
+`;
+
+const DELETE_CRYPTONEWS_FOR_CRYPTOCURRENCY = gql`
+    mutation DeleteCryptoNews($id: ID!) {
+        removeCryptoNews(id: $id) {
+            id
+            title
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Do we need the title of a removed news?

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 19]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
+    }
+`;
+
 export default function CryptoInformation(props: any) {
     const title = "Cryptocurrency information";
+    const [currentPage, setCurrentPage] = useState(1);
+    const [totalNumberOfCryptoNews, setTotalNumberOfCryptoNews] = useState(Number);
 
     const { id } = useParams<string>();
+    const currencyId = `http://purl.org/net/bel-epa/doacc#${id}`;
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Rename it to `cryptocurrencyId`

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 20]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
 
-    const { data, loading, error } = useQuery(GET_CRYPTOCURRENCY_BY_ID_QUERY, {
+    const { data: currencyData, loading: currencyLoading, error: currencyError } = useQuery(GET_CRYPTOCURRENCY_BY_ID_QUERY, {
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: `cryptocurrencyData`, `cryptocurrencyLoading`, `cryptocurrencyError`

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 21]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
         variables: {
             id: `http://purl.org/net/bel-epa/doacc#${id}`
-        }
+        },
+        context: {clientName: 'endpoint1'}
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Update the client name in order to reflect the previous comments

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 22]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
+    });
+
+    const getPaginatedNewsInput: GetPaginatedCryptoNewsInput = {
+        cryptocurrencyId: `http://purl.org/net/bel-epa/doacc#${id}`,
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Use the `cryptocurrencyId` defined above

File: cryk-ui/src/pages/CryptoInformation.tsx [hunk 23]
```diff
@@ -25,38 +28,126 @@ const GET_CRYPTOCURRENCY_BY_ID_QUERY = gql`
...
     useEffect(() => {
-        if (data) {
-            setCryptocurrency(data.cryptocurrency);
+        if (currencyData) {
+            setCryptocurrency(currencyData.cryptocurrency);
+        }
+    }, [currencyLoading, currencyData]);
+    useEffect(() => {
+        if (newsData) {
+            console.log(newsData.cryptoNews)
```
Comment (IoanSava) [inline cryk-ui/src/pages/CryptoInformation.tsx]: Remove this print

File: news-service/src/services/news-service.ts [hunk 24]
```diff
@@ -78,13 +81,25 @@ export const getCryptoNewsByCryptocurrencyId = async (cryptocurrencyId: string):
         }
     });
 
-    if (result["@graph"].length === 0) {
-        throw new Error(`No CryptoNews for Cryptocurrency with id ${cryptocurrencyId} found`);
```
Comment (IoanSava) [inline news-service/src/services/news-service.ts:82]: Oops, good catch

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+    context: { clientName: refetchInput.context },
+    refetchQueries: [{ query: refetchInput.query, context: { clientName: refetchInput.context }, variables: refetchInput.variables }]
+  });
+
+  const onSubmitHandler: SubmitHandler<NewsCardInput> = (values) => {
+    if (props.operationType === "create") {
+      const operationInput: CreateCryptoNewsInput = { title: values.title, body: values.body, about: [props.cryptocurrencyId] };
+      createUpdateNewsEntry({
+        variables: { createCryptoNewsInput: operationInput }
+      }).catch((error) => { console.log(JSON.stringify(error, null, 2)) });
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: nit: use `console.error`

File: cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+    }
+    else {
+      const operationInput: UpdateCryptoNewsInput = { id: (props.news ? props.news.id : ""), title: values.title, body: values.body };
+      createUpdateNewsEntry({
+        variables: { updateCryptoNewsInput: operationInput }
+      }).catch((error) => { console.log(JSON.stringify(error, null, 2)) });
```
Comment (IoanSava) [inline cryk-ui/src/components/CreateUpdateNewsCardDialog.tsx]: nit: use `console.error`

**Review by IoanSava — APPROVED** [review]

**Pull request merged**